### ISSUE TITLE
Migrate to Jetpack Compose (master)

### ISSUE THREAD

**Issue by westnordost** [streetcomplete/StreetComplete#5616]
This is the master ticket for migrating from the Android view system (layout XMLs etc.) to [Jetpack Compose](https://developer.android.com/jetpack/compose) as UI framework.

The primary reason for this is #5421: To make possible an iOS version of this app that has a shared codebase for the UI through the use of [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/). The first step however is to migrate to Jetpack Compose and only later switch to Compose Multiplatform.

The first foray into Jetpack Compose has been made in PR #5447, I am also new to this but I tried to follow guidelines closely how things should be done, so feel free to look into that PR to learn how the UI code typically looks like before and after.

## Contributing

As this is a larger undertaking which can be implemented step by step - this ticket shall serve as a master ticket. Whenever you would like to contribute, post on which screen or view you would like to work on now here and later post a PR when you are done. 

Migration is generally done bottom-up, i.e. start with single views first and then work your way up to the screen-level, see [Migration Strategy](https://developer.android.com/jetpack/compose/migrate/strategy). For converting a whole screen, it is very advisable to do this only after the Fragment or Activity in question has been refactored to use ViewModels, see #5530.

**Comment by westnordost** [comment]
**Current blockers and issues with Compose:**

- Appear-animation missing for achievements and links screen. [Lazy item animations not supported by Jetpack Compose yet](https://developer.android.com/jetpack/androidx/compose-roadmap).

- Fancy animate-from-icon animation is missing for achievement dialog. [Not supported by Jetpack Compose yet](https://developer.android.com/develop/ui/compose/animation/shared-elements)

- (Material2) DatePicker and TimePicker (dialogs) are not available in Jetpack Compose (yet) 

- scroll bars are not supported by Jetpack Compose yet. (They would be handy at least in the Log screen)

**Comment by westnordost** [comment]
Re-do of About and Settings screen done, in review. Details here: https://github.com/streetcomplete/StreetComplete/pull/5719/files

**Comment by westnordost** [comment]
User screens are also done, except for the quest statistics / country statistics screen. See #5595

**Comment by westnordost** [comment]
I'll look into the two tutorial screens (start tutorial, overlay tutorial) coming week.

**Comment by westnordost** [comment]
tutorials screen is done

**Comment by westnordost** [comment]
Next up would be the buttons and controls on the main screen plus the menus and dialogs there.

So, this is an opportunity to redesign the main menu dialog, which I designed the way it is now chiefly because it was the easiest to do. In the future it could e.g. also be a popup menu, or a sheet that slides in (from the top, from the right?) like the undo menu does... I am open to suggestions. What's popular in other apps?

**Comment by mnalis** [comment]
Other approaches I see:

- slide from the left - e.g. Osmand (which has menu icon on lower left) and OpenStop, K-9, Libretorrent, Nextcloud, AntennaPo, NewPipe, Fedilab (which have menu icon on upper left corner)
- slide from the bottom - e.g. Commons App, Vespucci, Organic Maps (which have menu icon on lower right)
- popup fullscreen - e.g. EveryDoor (which has menu icon on upper left), F-droid (lower right)

I actually quite like current StreetComplete _"pop up in middle of the screen"_ menu, but other approaches are probably fine too, provided they happen fast (i.e. noticeably faster than half a second).

**Comment by westnordost** [comment]
Thanks for the input! I ended up mirroring the current appearance now.

I am now working on migrating the UI of the main screen to compose, i.e.
- (button) controls and star counter, (dropdown on long-press)
- main menu
- team mode menu (I plan to redesign it to be a wizard, i.e. similar appearance as the tutorial screens and will commission an artist for some nice illustration)
- overlay selection dropdown
- the various messages-dialogs (new-osm-message, "too much to do"-message, achievement unlocked, changelog)
- upload/download-related dialogs probably
- **location pointer pin**: **However, I have no idea how to implement that with compose! Maybe this feature has to go! :-(** The problem is that the layout of the location pointer pin depends on the layout of the other elements, i.e. it has to be layout after the bounds of the other buttons are known - but in Compose, it is not really possible to inspect that! **_Help appreciated!_** 


Maybe now, maybe immediately afterwards:
- edit history sidebar and dialog

**Comment by rugk** [comment]
> slide from the bottom 

…generally has the advantage [so they can be used with one finger](https://medium.com/@uxandyouti/thumb-friendly-design-optimizing-mobile-ui-for-one-handed-use-0f4acc446b3f). And [IMHO such a panel](https://m3.material.io/foundations/interaction/gestures#22462fb2-fbe8-4e0c-b3e7-9278bd18ea0d) optionally including a gesture even (moving up from bottom) would be good too in SC, I don't quite like that you e.g. need to press (/go to) the top right icon for opening the menu.

**Comment by rugk** [comment]
By that migration, BTW; maybe the app could or will automatically get more accessible BTW (https://m3.material.io/foundations/interaction/gestures#22462fb2-fbe8-4e0c-b3e7-9278bd18ea0d). At least Kotline Multiplatorm [has a page for that, already](https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-ios-accessibility.html#customize-synchronization-of-the-accessibility-tree) and I know iOS devices are quite popular for blind or visually impaired people. Well… let's see. (Just came upon this idea [because another Protoype fund project currently offers free a10y reviews](https://github.com/KreerC/a11ybuddy)).

**Comment by westnordost** [comment]
the complete user screen is now based on compose (including the statistics screen)

**Comment by westnordost** [comment]
The main screen, i.e. the buttons, the menus, the dialogs, the team mode, the dropdowns, the edit history sidebar, error handling and handling start parameters is also all based on compose. See #5799 (currently in alpha)

**Comment by westnordost** [comment]
Next step would be

- the map, but this depends on a library that does not exist yet.
- and migrate all those widgets used in quest forms to compose: the street side select widget, the "select one of these icons+text" widget, the opening hours editing, the names editing etc.. These are many small(er) tasks, so are well eligible for individual contributions for PRs. Write here if you are interested.

**Comment by GaeaKat** [comment]
I have a bit of a silly question. With the original being ![Image](https://github.com/user-attachments/assets/054c8350-d438-4790-9216-7ee7a4c08148) of course, is
![Image](https://github.com/user-attachments/assets/a4b0225f-efe7-4b7d-8add-b648e8bd2214)
 this a suitable match, or would you prefer more exact match?

**Comment by westnordost** [comment]
Oh cool, you are contributing!

It doesn't need to exactly look the same. It's more important to rather use mostly the normal (material) components than to endlessly customize the whole thing. (I.e. better to have less code than to have a perfect match with how it looks like currently)

**Comment by GaeaKat** [comment]
@westnordost Do you prefer smaller PR's with single items, like, a PR for the button, then PR for another widget, or to let them build into a bigger PR?

**Comment by westnordost** [comment]
I prefer small PRs which however form a logical unit.

E.g. a PR for just one of those buttons when they are not used anywhere yet would be a bit too small. The apply-last-answers-button-row plus building+roof level input form would be ideal as a PR. The apply-last-answers-button-row could maybe be generalized to be easily reusable in other forms with such an UI element, but work on other widgets should ideally go into a follow-up PR.
Since anyway this is your first PR, better keep it small and manageable as depending on your experience level (in Compose, in OSM, in StreetComplete) I may have comments about this or that.

**Comment by GaeaKat** [comment]
I have rapidly found that what I thought I knew about compose was not enough to do much more then make the button, and wire it into the existing form currently.  I am going to be spending a week or two learning / playing with less complicated projects so I can continue attempting to contribute to this project. in the mean time, will submit a draft PR so not just dead in water

**Comment by GaeaKat** [comment]
https://github.com/streetcomplete/StreetComplete/pull/6022 this is it wired up, and working in the debug menu!

**Comment by westnordost** [comment]
**Note**: When the note comments display is migrated to Jetpack Compose, the text from the notes should be formatted with a spacing of 1.5 lines on every (explicit) line break to have more click-space for e.g. links to attached images. See #6053

**Comment by westnordost** [comment]
An important big step towards migrating to Compose is #6072

**Comment by westnordost** [comment]
Because @Ercalvez asked in another ticket, here's a (not necessarily complete) list of UI of quest forms that need migration to Compose. As mentioned earlier, #6022 is a good blueprint how the migration would look like.

The animal icon is a rough estimation how big it would be. 🐁 < 🐈‍⬛ < 🐎 < 🐘 < 🐋

### done
- 🐁 ~autocomplete suggestions input (e.g. clothes containers operator)~ ([#6603](https://github.com/streetcomplete/StreetComplete/pull/6603)) @westnordost
- 🐁 ~level select (e.g. level of shop)~ (https://github.com/streetcomplete/StreetComplete/commit/f6999d5f83638892c00f800fb52d0e2e0d9f3dab) @westnordost
- 🐁 ~count input~ ([#6282](https://github.com/streetcomplete/StreetComplete/pull/6282)) @Ercalvez 
- 🐁 ~location input (e.g. location of defibrillator)~ (https://github.com/streetcomplete/StreetComplete/commit/e5345907f86647a0ad42104be47c07ae37d15dfb) @westnordost
- 🐁 ~ref input (e.g. hydrant ref)~ (https://github.com/streetcomplete/StreetComplete/commit/64126c87fa9c507b7fc21ed7464a1b09783c7888) @westnordost
- 🐈‍⬛ ~fire hydrant diameter~ ([#6395](https://github.com/streetcomplete/StreetComplete/pull/6395)) @westnordost 
- 🐈‍⬛ ~building levels~ ([#6022](https://github.com/streetcomplete/StreetComplete/pull/6022)) @GaeaKat 
- 🐎 ~width & height inputs~ ([#6285](https://github.com/streetcomplete/StreetComplete/pull/6285)) @ltoenning 
- 🐎 ~max weight~ ([#6477](https://github.com/streetcomplete/StreetComplete/pull/6477)) @westnordost 
- 🐎 ~text-based list~ ([#6249](https://github.com/streetcomplete/StreetComplete/pull/6249)) @GaeaKat 
- 🐎 ~housenumber input~ ([#6403](https://github.com/streetcomplete/StreetComplete/pull/6403)) @westnordost 
- 🐎 ~parking fee (including "time restriction" widget)~ ([#6691](https://github.com/streetcomplete/StreetComplete/pull/6691)) @westnordost 
- 🐘 ~lanes select~ ([#6490](https://github.com/streetcomplete/StreetComplete/pull/6490)) @westnordost 
- 🐘 ~list of localized names (street name, place name, ...), name with suggestions~ ([#6403](https://github.com/streetcomplete/StreetComplete/pull/6403)) @westnordost 
- 🐘 ~opening hours~ ([#6691](https://github.com/streetcomplete/StreetComplete/pull/6691)) @westnordost 
- 🐋 ~image+text-based list~ [#6490](https://github.com/streetcomplete/StreetComplete/pull/6490) @westnordost 
- 🐋 ~street side select (bike paths, parking, sidewalks, ...)~ ([#6490](https://github.com/streetcomplete/StreetComplete/pull/6490)) @westnordost 
- 🐘 ~max speed~ @westnordost  ([#6758](https://github.com/streetcomplete/StreetComplete/pull/6758))
- 🐘 ~POI selection (shop type, selecting new POI when old doesn't exist anymore, ...)~ ([#6773](https://github.com/streetcomplete/StreetComplete/pull/6773)) @westnordost
- 🐎 ~note discussion + attach photo~ ([#6804](https://github.com/streetcomplete/StreetComplete/pull/6797)) @westnordost
- 🐁 ~move node form~ [#6797](https://github.com/streetcomplete/StreetComplete/pull/6797) @michaelabon

### to do
- 🐈‍⬛ split way form

**Comment by westnordost** [comment]
Will probably do POI selection or opening hours next.

**Comment by westnordost** [comment]
Opening hours + collection times + parking fee with maxstay etc. done. Next, I'll do max speed.

**Comment by westnordost** [comment]
Aaand, the last quest forms: POI selection for things and places overlay and the dialog that asks "what's here now?"

**Comment by westnordost** [comment]
Currently on note forms + attach photo

**Comment by riQQ** [comment]
For reference:
- #6804

**Comment by westnordost** [comment]
So, what follows now is to migrate the view that surrounds the quest form, i.e. the view that contains the buttons at the bottom, the Uh... menu, the floating OK button, the quest title, subtitle + hint area and user notes bubble.

This then allows (actually, requires) to also finally migrate the remaining dialogs that pop up as a result of e.g. tapping an "other answer" option to compose.

This will be a really big PR, and I reserve this to do myself. 

Preparatory work:
- ~migrate quest title + subtitle + hint area to compose~ ([#6784](https://github.com/streetcomplete/StreetComplete/pull/6784)) @westnordost 
- create layout for this in compose ([#6818](https://github.com/streetcomplete/StreetComplete/pull/6818)) @westnordost 
- migrate form-portion of move node, split way, note creation, ...

**Comment by riQQ** [comment]
- #6842

**Comment by westnordost** [comment]
Alright, I am happy to announce that with #6842 merged, this is pretty much done. For the migration to Maplibre-compose, we already have a separate issue/pr (#6352) and otherwise there are only a few dialogs left which will be eliminated in the frame of other tickets.

🥳
